core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -174,26 +170,33 @@ public boolean allocateUnassigned(RoutingAllocation allocation, long allocateUna
                 }
             } else if (matchingNodes.hasAnyData() == false) {
                 // if we didn't manage to find *any* data (regardless of matching sizes), check if the allocation
-                // of the replica shard needs to be delayed, and if so, add it to the ignore unassigned list
+                // of the replica shard needs to be delayed, and if so, compute the delay and add it to the ignore unassigned list
                 // note: we only care about replica in delayed allocation, since if we have an unassigned primary it
                 //       will anyhow wait to find an existing copy of the shard to be allocated
                 // note: the other side of the equation is scheduling a reroute in a timely manner, which happens in the RoutingService
-                IndexMetaData indexMetaData = allocation.metaData().index(shard.getIndex());
-                long delay = shard.unassignedInfo().getDelayAllocationExpirationIn(allocateUnassignedTimestapm, settings, indexMetaData.getSettings());
-                if (delay > 0) {
-                    logger.debug("[{}][{}]: delaying allocation of [{}] for [{}]", shard.index(), shard.id(), shard, TimeValue.timeValueMillis(delay));
-                    /**
-                     * mark it as changed, since we want to kick a publishing to schedule future allocation,
-                     * see {@link org.elasticsearch.cluster.routing.RoutingService#clusterChanged(ClusterChangedEvent)}).
-                     */
-                    changed = true;
-                    unassignedIterator.removeAndIgnore();
-                }
+                changed |= ignoreUnassignedIfDelayed(System.nanoTime(), allocation, unassignedIterator, shard);
             }
         }
         return changed;
     }
 
+    // public for tests
+    public boolean ignoreUnassignedIfDelayed(long timeNowNanos, RoutingAllocation allocation, RoutingNodes.UnassignedShards.UnassignedIterator unassignedIterator, ShardRouting shard) {
```

Comment:
I also think we could now make this static and unit test *just* this method, which would be really helpful